crates/symbolicator-service/src/utils/http.rs

```diff
@@ -47,6 +49,126 @@ pub fn create_client(config: &Config, trusted: bool) -> reqwest::Client {
     builder.build().unwrap()
 }
 
+#[derive(Debug, Clone)]
+struct AllowedOriginPattern {
+    scheme: String,
+    domain: String,
+    path: String,
+}
+
+impl AllowedOriginPattern {
+    fn parse(value: &str) -> Option<Self> {
+        let (scheme, rest) = value.split_once("://").unwrap_or(("*", value));
+        let (domain, path) = rest.split_once('/').unwrap_or((rest, "*"));
+
+        let (domain, port) = match rest.split_once(':') {
+            Some((d, p)) => (d, Some(p)),
+            None => (domain, None),
+        };
+
+        let mut domain = idna::domain_to_ascii(domain).ok()?;
+
+        if let Some(port) = port {
+            write!(&mut domain, ":{port}").unwrap();
+        }
+
+        Some(Self {
+            scheme: scheme.to_string(),
+            domain,
+            path: path.to_string(),
+        })
+    }
+}
+
+/// Checks if the given `origin` matches one of the patterns in the `allowed` list.
+///
+/// Allowed origins may be defined in several ways:
+/// - `http://domain.com[:port]`: Exact match for base URI (must include port).
+/// - `*`: Allow any domain.
+/// - `*.domain.com`: Matches domain.com and all subdomains, on any port.
+/// - `domain.com`: Matches domain.com on any port.
+/// - `*:port`: Wildcard on hostname, but explicit match on port.
+///
+/// This transparently handles [IDNA-encoded domains](https://en.wikipedia.org/wiki/Internationalized_domain_name).
+pub fn is_valid_origin(origin: &str, allowed: &[&str]) -> bool {
+    if allowed.is_empty() {
+        return false;
+    }
+
+    if allowed.contains(&"*") {
+        return true;
+    }
+
+    let origin = origin.to_lowercase();
+
+    // Fast check
+    if allowed.contains(&&origin[..]) {
+        return true;
+    }
+
+    // XXX: In some cases origin might be localhost (or something similar) which causes a string value
+    // of 'null' to be sent as the origin
+    if origin == "null" {
```

Comment:
It's gone in the current version. The `origin` parameter is now a `Url`, since at the point where we use this, we definitely have an actual URL, which makes this irrelevant.